src/main/scala/com/fulcrumgenomics/bam/UpdateReadGroups.scala

```diff
@@ -0,0 +1,201 @@
+      updateRecord(record, outHeader, toReadGroupsMap)
+      out.addAlignment(record)
+      progress.record(record)
+    }
+    logger.info(s"Processed ${progress.getCount} in ${progress.getElapsedSeconds} seconds")
+
+    CloserUtil.close(in)
+    out.close()
+  }
+
+  private def updateRecord(record: SAMRecord,
+                           header: SAMFileHeader,
+                           readGroupMap: Map[String, SAMReadGroupRecord]): SAMRecord = {
+    val fromReadGroup = record.getReadGroup match {
+      case null => throw new IllegalStateException("Record '${record.getReadName}' has no read group")
+      case rg => rg
+    }
+    readGroupMap.get(fromReadGroup.getId) match {
+      case None =>
+        throw new IllegalStateException(s"Read group id '${record.getReadGroup.getId}' from record '${record.getReadName}' not found in the SAM header.")
+      case Some(readGroup: SAMReadGroupRecord) =>
+        record.setAttribute(SAMTag.RG.name(), readGroup.getId)
+        record.setHeader(header)
+    }
+    record
+  }
+
+  private def mappingFileError(msg: String, lineNumber: Int): Nothing = {
+    throw new IllegalStateException(s"Mapping file line $lineNumber: $msg")
+  }
+
+  private def getReadGroupValue(key: String, header: Seq[String], values: Seq[String], lineNumber: Int): String = {
+    val idx = header.indexOf(key)
+    if (idx == -1) throw new IllegalArgumentException(s"Key '$key' not found in the mapping file header: " + header.mkString(","))
+    if (values.size <= idx) throw new IllegalArgumentException(s"Line $lineNumber has missing values in the mapping file")
+    values(idx)
+  }
+
+  private def getReadGroupMap(readGroupsFile: FilePath, samFileHeader: SAMFileHeader): Map[String, SAMReadGroupRecord] = {
+    // create the new read groups
+    val fromReadGroups = samFileHeader.getReadGroups.map { rg => rg.getId -> rg }.toMap
+
+    // read in the lines
+    val lines = Io.readLines(readGroupsFile).map(_.trim).toSeq
+    if (lines.size <= 1) mappingFileError("No read groups to replace were given in the mapping file.", 1)
+
+    // parse in the header
+    val header = lines.head.split(",")
+    // make sure we recognize all values in the header
+    header.foreach { key =>
+      if (key != fromReadGroupKey && !SAMReadGroupRecord.STANDARD_TAGS.contains(key)) mappingFileError(s"Read group key '$key' not recognized.", 1)
+    }
+    // make sure there are no duplicates
+    val headerKeySet = header.toSet
+    header.find(!headerKeySet.contains(_)) match {
+      case None =>
+      case Some(key) => mappingFileError(s"Duplicate read group key '$key'", 1)
+    }
+    // make sure there is at least the id
+    if (!header.contains(SAMReadGroupRecord.READ_GROUP_ID_TAG)) mappingFileError(s"Missing header key '${SAMReadGroupRecord.READ_GROUP_ID_TAG}'", 1)
+
+    // parse in the rest of the lines
+    val readGroupMaps = lines.tail.zipWithIndex.map { case (line: String, idx: Int) =>
+      line.split(",").toList match {
```

Comment:
I think technically `split(String)` is by regex, and `split(Char)` is by character, so you might want to use `split(',')`